rickroll.py

```diff
@@ -153,5 +174,55 @@ def original():
 
     return str(play_tune(tune))
 
+@app.route("/sms", methods=["POST"])
+def sms():
+    """Adds an incoming SMS to a queue, to reply to later."""
+    bucket.put_object(
+        Key='queue/{to}/{from_}'.format(to=request.form['To'],
+                                        from_=request.form['From']),
+        Body='',
+    )
+    return "Hello world!"
+
+def send_sms():
+    """Empties the queue of incoming SMSes, replying to each one."""
+    for queue_entry in bucket.objects.filter(Prefix='queue/'):
+        _, our_number, their_number = queue_entry.key.split("/")
+        state_key = "state/{}/{}".format(our_number, their_number)
+
+        # Find out which message we sent last, if any
+        # Error handling taken from https://stackoverflow.com/a/33843019
+        try:
+            state_obj = s3.Object(
+                os.environ['data_bucket'],
+                state_key,
+            )
+            state_obj.load()
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                last_msg = -1
+            else:
+                raise
+        else:
+            last_msg = int(state_obj.get()['Body'].read().decode('utf8'))
+
+        # Send the next one
+        next_msg = last_msg + 1
+        if next_msg < len(messages):
+            twilio_client.messages.create(
+                to=their_number,
+                from_=our_number,
+                body=messages[next_msg],
+            )
+
+        # Record which message we just sent in S3
+        bucket.put_object(
+            Key=state_key,
+            Body=str(next_msg),
+            Expires=datetime.now() + timedelta(days=7),
```

Comment:
Correct. I wasn't sure what the threshold should be, but I felt like there should be _some_ cut-off.